p2p/peer.go

```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
+	}
+
+	ranked, err := m.store.Ranked()
 	if err != nil {
 		return "", PeerAddress{}, err
 	}
-	for _, peer := range peers {
-		switch {
-		case len(peer.Addresses) == 0:
-		case m.dialing[peer.ID]:
-		case m.connected[peer.ID]:
-		default:
-			// FIXME: We currently only dial the first address, but we should
-			// track connection statistics for each address and return the most
-			// appropriate one.
+	for _, peer := range ranked {
+		if m.dialing[peer.ID] || m.connected[peer.ID] {
+			continue
+		}
+
+		for _, addressInfo := range peer.AddressInfo {
+			if time.Since(addressInfo.LastDialFailure) < m.retryDelay(peer, addressInfo.DialFailures) {
+				continue
+			}
+
+			// At this point we have an eligible address to dial. If we're full
+			// but have peer upgrade capacity (as checked above), we need to
+			// make sure there exists an evictable peer of a lower score that we
+			// can replace. If so, we can go ahead and dial this peer, and
+			// EvictNext() will evict a lower-scored one later.
+			//
+			// If we don't find one, there is no point in trying additional
+			// peers, since they will all have the same or lower score than this
+			// peer (since they're ordered by score via peerStore.Ranked).
+			//
+			// FIXME: There is a race condition here where, if there exists a
+			// single lower-scored peer, we may end up dialing multiple
+			// higher-scored new peers that all expect the same lower-scored
+			// peer to be evicted, causing us to take on too many peers. We may
+			// need to reserve the eviction for this specific peer such that
+			// others can't claim it.
+			if m.options.MaxConnected > 0 &&
+				len(m.connected) >= int(m.options.MaxConnected) &&
+				!m.peerIsUpgrade(peer, ranked) {
+				return "", PeerAddress{}, nil
+			}
+
 			m.dialing[peer.ID] = true
-			return peer.ID, peer.Addresses[0], nil
+			return peer.ID, addressInfo.Address, nil
 		}
 	}
 	return "", PeerAddress{}, nil
 }
 
+// retryDelay calculates a dial retry delay using exponential backoff, based on
+// retry settings in PeerManagerOptions. If MinRetryTime is 0, this returns
+// MaxInt64 (i.e. an infinite retry delay, effectively disabling retries).
+func (m *PeerManager) retryDelay(peer *peerInfo, failures uint32) time.Duration {
+	if failures == 0 {
+		return 0
+	}
+	if m.options.MinRetryTime == 0 {
+		return time.Duration(math.MaxInt64)
+	}
+	delay := m.options.MinRetryTime * time.Duration(math.Pow(2, float64(failures)))
+	if peer.Persistent && m.options.MaxRetryTimePersistent > 0 {
+		// We have to check this separately, since we don't want to fall back to
+		// MaxRetryTime in cases where it is smaller than MaxRetryTimePersistent.
+		if delay > m.options.MaxRetryTimePersistent {
+			delay = m.options.MaxRetryTimePersistent
+		}
+	} else if m.options.MaxRetryTime > 0 && delay > m.options.MaxRetryTime {
+		delay = m.options.MaxRetryTime
+	}
+	delay += time.Duration(rand.Int63n(int64(m.options.FuzzRetryTime))) // nolint:gosec
```

Comment:
Should we not create a PRNG object in the manager with a seed (i.e. time.now())?

Reply:
Yes, once this is no longer a prototype. This would be needed for testing as well.